editor/src/components/canvas/canvas-strategies/post-action-options/post-action-paste.ts

```diff
@@ -159,28 +170,75 @@ function pasteChoiceCommon(
         }
       })()
 
+      const pathAfterReparent = elementPathFromInsertionPath(
+        target.parentPath,
+        elementWithUID.value.uid,
+      )
+
+      const originalPaths = pathPartsFromJSXElementChild(elementPaste.element, []).map((part) =>
+        appendPathPart(elementPaste.originalElementPath, part),
+      )
+
+      const pathsAfterUIDFix = pathPartsFromJSXElementChild(elementWithUID.value, []).map((part) =>
+        appendPathPart(pathAfterReparent, part),
+      )
+
+      const mapping: OldPathToNewPathMapping = zip(
+        originalPaths,
+        pathsAfterUIDFix,
+        (oldPath, newPath) => ({ oldPath, newPath }),
+      ).reduce(
+        (accI: OldPathToNewPathMapping, { oldPath, newPath }) => ({
+          ...accI,
+          [EP.toString(oldPath)]: newPath,
+        }),
+        {},
+      )
+
+      oldPathToNewPathMapping = {
+        ...oldPathToNewPathMapping,
+        ...mapping,
+      }
+
       return {
         elementPath: elementPaste.originalElementPath,
         pathToReparent: elementToReparent(elementWithUID.value, elementPaste.importsToAdd),
         intendedCoordinates: intendedCoordinates,
-        uid: elementWithUID.value.uid,
+        newUID: elementWithUID.value.uid,
       }
     })
+
   return staticReparentAndUpdatePosition(
     target,
     editorStateContext,
     pasteContext,
     elementsToInsert,
     indexPosition,
+    oldPathToNewPathMapping,
   )
 }
 
+function appendPathPart(path: ElementPath, part: PathPart): ElementPath {
```

Comment:
This could do with a little explanatory comment to indicate why it needs using over the similarly named functions it invokes.